src/services/mediaDevicePreferences.ts

```diff
@@ -0,0 +1,182 @@
+	const newVideoInputList = videoInputList.slice()
+
+	for (const device of devices) {
+		if (device.kind === 'audioinput') {
+			// Add to the list of known devices
+			if (device.deviceId !== 'default' && !newAudioInputList.some(input => input.deviceId === device.deviceId)) {
+				registerNewDevice(device, newAudioInputList)
+				audioHasChanged = true
+			}
+		} else if (device.kind === 'videoinput') {
+			// Add to the list of known devices
+			if (device.deviceId !== 'default' && !newVideoInputList.some(input => input.deviceId === device.deviceId)) {
+				registerNewDevice(device, newVideoInputList)
+				videoHasChanged = true
+			}
+		}
+	}
+
+	return {
+		newAudioInputList: audioHasChanged ? newAudioInputList : null,
+		newVideoInputList: videoHasChanged ? newVideoInputList : null,
+	}
+}
+
+/**
+ * Update devices preferences. Assuming that preferred devices were selected, should be called after applying the selection:
+ * so either with joining the call or changing device during the call
+ * Regardless if new unknown or registered device was selected, we promote it in the list:
+ * - if first item is plugged, then we prefer device out of all options and put it on the first place;
+ * - if first item is unplugged, then we don't consider it, and compare with the next in the list;
+ * - if second item is unplugged, compare with next in the list;
+ * - ...
+ * - if N-th item is plugged, then we prefer device to it and put it on the Nth place.
+ *
+ * Returns changed preference lists for audio / video devices (null, if it hasn't been changed)
+ *
+ * @param devices list of available devices
+ * @param audioInputId id of currently selected audio input
+ * @param videoInputId id of currently selected video input
+ * @param audioInputList list of registered audio devices in order of preference
+ * @param videoInputList list of registered video devices in order of preference
+ */
+function updateMediaDevicesPreferences(devices: MediaDeviceInfo[], audioInputId: string, videoInputId: string, audioInputList: MediaDeviceInfo[], videoInputList: MediaDeviceInfo[]) {
+	let audioHasChanged = false
+	let videoHasChanged = false
+	const newAudioInputList = audioInputList.slice()
+	const newVideoInputList = videoInputList.slice()
+	const availableDevices = devices.map(device => device.deviceId).filter(id => id !== 'default')
+
+	const audioDeviceRank = newAudioInputList.findIndex(device => device.deviceId === audioInputId)
+	if (audioDeviceRank === -1 && audioInputId !== 'default' && audioInputId !== null) {
+		const device = devices.find(device => device.deviceId === audioInputId)
+		if (device) {
+			console.debug('Registering new audio device:', device)
+			newAudioInputList.unshift(device)
+			audioHasChanged = true
+		}
+	} else if (audioDeviceRank > 0) {
+		const device = newAudioInputList.splice(audioDeviceRank, 1)[0]
+		const pluggedIndex = newAudioInputList.findIndex(device => availableDevices.includes(device.deviceId))
```

Comment:
```suggestion
		// Remove the device from the list
		const device = newAudioInputList.splice(audioDeviceRank, 1)[0]
		// Get the index of the first plugged device
		const pluggedIndex = newAudioInputList.findIndex(device => availableDevices.includes(device.deviceId))
```

let's add some comments in this update logic

Reply:
Was about to rewrite that logic